source/ConversationPanel.cpp

```diff
@@ -340,38 +342,70 @@ void ConversationPanel::Goto(int index, int selectedChoice)
 	// We'll need to reload the choices from whatever new node we arrive at.
 	choices.clear();
 	node = index;
-	// Not every conversation node allows a choice. Move forward through the
-	// nodes until we encounter one that does, or the conversation ends.
-	while(node >= 0 && !conversation.IsChoice(node))
+	while(node >= 0)
 	{
-		int choice = 0;
-		if(conversation.IsBranch(node))
+		// Not every conversation node allows a choice. Move forward through the
+		// nodes until we encounter one that does, or the conversation ends.
+		while(node >= 0 && !conversation.IsChoice(node))
```

Comment:
I feel like we should be able to avoid indenting this whole block, but I haven't truly checked that.

Reply:
The inner loop scans for the first choice node. Then, conditions are checked, and if every choice node was skipped because of conditions, the outer loop iterates again.

I'm not sure if there's a clearer way to represent the desired code flow here. It's a bit of a weird case.

Reply:
Good to check such corner-cases; we don't want the conversation to get stuck because the player has to choose between zero choices and cannot select one.
A possibly cleaner way to handle this loop is to update the `IsChoice()` function to check if there is at least one valid choice, then we might not need the double loop.

Reply:
The reason to have a double loop is for conversations structured like:

- Say thing 1
- Iff condition A, offer choice 1A
- Iff condition B, offer choice 1B
- Say thing 2
- Iff condition C, offer choice 2C
- Iff condition D, offer choice 2D
- Say thing 3

so that, if none of A, B, C, or D, then it acts just like:

- Say thing 1
- Say thing 2
- Say thing 3

Reply:
Yes, if none of A or B, then the amount of choices after "Say thing 1" are zero. And if none of C or D, then the amount of choices after "Say thing 2" are zero. So I expect my proposal to work just as you specify here, because IsChoice() would return zero if no valid choice is found.

Reply:
Maybe call the new function `HasAnyChoice()`?

Reply:
I like that name too.

If I'm able to revisit this in detail, I'll try it out with the suggested change. It seems like it should work.

Reply:
When I was implementing all the requested changes in the reviewer comments, I couldn't figure out what you were requesting here. The only effects I see of moving the loop out to a function is to increase the code length and reduce indentation. In my mind, one level of indentation doesn't matter, while longer code makes it harder to maintain. I don't have a strong feeling on these matters, I'm just trying to parse all of your comments.

I'm 93% certain I don't understand what you want, but if you can explain it to me, I can implement it.

Reply:
SolraBizna introduced an additional `while(node >= 0)` loop at line 356. This additional loop was needed to deal with the situation where a player has choices, but all choices are hidden (due to all the choices having conditions, and the conditions hiding all possible choices). I was suggesting to not add the additional loop (so remove the loop now at line 356), but to call a function that checks if at least one choice can be chosen in the other loop (currently at line 360).

Reply:
Are you saying we should have a single loop with something like this?

```
while(node >= 0 && (!conversation.IsChoice(node) || !conversation.HasAnyChoice(node)))
```

Reply:
That change will simplify the code, so I see why you want it.

I guess HasAnyChoice() can call IsChoice(), letting you only have two conditions in the while().

Reply:
I implemented it, and put it in a PR to this branch here: https://github.com/SolraBizna/endless-sky/pull/3

Reply:
I had to handle a corner case. A choice with zero options is how the game represents "set the player's name." I had to differentiate between a choice with no options, and a choice with options, none of which are displayed.